schema.sql: Add more constraints

RyanSquared opened the pull request:
Added the following constraints:
- check_hosts_for_hosts(): Check whether hosts have too many users
  before changing the maximum user limit
- check_invalid_name_for_passwd(): Checks whether a username exists in
  a group before inserting into passwd
- check_invalid_name_for_group(): Checks whether a group name exists in
  a passwd entry before inserting into group
- check_banned_username(): Check whether a username has been banned,
  such as if it's a name reserved for sysadmin usernames

Fixes: Part of #17, #27, #28.

daurnimator commented on schema.sql:
```diff
@@ -62,6 +75,49 @@ create constraint trigger max_users
+create constraint trigger check_name_exists_group
+    after insert or update on "group"
+    for each row
+    execute procedure check_invalid_name_for_group();
```
Could still have `when (new.name <> old.name)`

RyanSquared replied:
Only for update and I don't think we would ever update enough to be relevant

daurnimator commented on schema.sql:
```diff
@@ -62,6 +75,49 @@ create constraint trigger max_users
+    execute procedure check_invalid_name_for_group();
+
+-- prevent users from taking names typically reserved for sysadmins
+CREATE TABLE "banned_usernames" (
```
different case to the rest.

RyanSquared replied:
I followed the case used when creating other tables.

daurnimator replied:
oh. we should fix them all at once then.

daurnimator commented:
>```
> ERROR:  INSERT trigger's WHEN condition cannot reference OLD values at character 116
> STATEMENT:  create constraint trigger check_name_exists_passwd
> 	    after insert or update on passwd
> 	    for each row
> 	    when (old.name <> new.name)
> 	    execute procedure check_invalid_name_for_passwd();
> psql:schema.sql:99: ERROR:  INSERT trigger's WHEN condition cannot reference OLD values
> LINE 4:     when (old.name <> new.name)
>                   ^
> ```

Should do: `tg_op = 'INSERT' or old.name <> new.name`

RyanSquared commented:
`tg_op` doesn't exist at the time of `when`, logic will have to go in the function

RyanSquared commented:
Reverted commit due to `tg_op` not existing at the point of `when`.

daurnimator commented:
Don't revert in a PR. Just remove the commit.

RyanSquared commented:
I figure leave the commits in the PR if someone else makes the comment, then rebase and drop before merging. But these are enough comments that no context is lost.

RyanSquared commented:
I'd like another review, and for someone else to be the one to merge _because I don't actually have write access_.

RyanSquared commented:
@daurnimator feel free to merge whenever if no one else wants to review it.

KellerFuchs commented:
@RyanSquared I rebased this, since it was conflicting with other stuff that got merged (#32 I guess).
I started reviewing a few days ago, but didn't have spare brains much; will try to finish by next weekend.

KellerFuchs commented on schema.sql:
```diff
@@ -62,6 +75,49 @@ create constraint trigger max_users
+    execute procedure check_invalid_name_for_group();
+
+-- prevent users from taking names typically reserved for sysadmins
+CREATE TABLE "banned_usernames" (
```
`reserved_usernames`?  `banned_usernames` sounds like is it for users who misbehaved.

KellerFuchs replied:
Also, should the table get populated?

RyanSquared replied:
yeah, reserved makes more sense. Also, I seemed to have forgotten to push it, but I had a thing in the Makefile to insert a bunch of names from the list on the issue relating to this section. Will do later.

KellerFuchs replied:
I renamed the table while splititng stuff into commits.

KellerFuchs commented on schema.sql:
```diff
@@ -62,6 +75,49 @@ create constraint trigger max_users
+        return new;
+    end $$;
+create constraint trigger check_banned_username
+    after insert on passwd
```
You also need to check after an insert in `banned_usernames`

RyanSquared replied:
Why? It already checks with the `UNIQUE` constraint whether or not it already exists.

KellerFuchs replied:
No, I meant that when inserting in `banned_usernames`, we should check that the username doesn't already exist.

RyanSquared replied:
Oh, alright, will fix tomorrow hopefully?

KellerFuchs replied:
I just fixed it

KellerFuchs reviewed (changes requested):
Great work, thanks!

I left some actionable feedback already, and I suspect there might be refactor opportunities, but it's 2.40am and I need to snooze.

KellerFuchs commented:
I forgot to mention in the review message, but all this needs tests.

KellerFuchs commented:
BTW, I would have split that in individual commits, so it's easier to find which bit closed #27 for instance.

RyanSquared commented:
Yeah, sorry, this was a bit weird of a PR. I basically wrote the code in the issue, realized afterwards "hey it actually works why not shove it all in a PR".

KellerFuchs reviewed (dismissed):
OK, there is still stuff that needs doing after I took another pass at it:
- [ ] actually populate the `reserved_usernames` table;
- [ ] add tests (positive and negative) for all those constraints.

KellerFuchs commented:
@daurnimator Can you have a look through my changes?

daurnimator commented on schema.sql:
```diff
@@ -62,6 +77,64 @@ create constraint trigger max_users
+    execute procedure check_invalid_name_for_group();
+
+-- prevent users from taking names which are typically reserved
+CREATE TABLE "reserved_usernames" (
```
case.

daurnimator commented on schema.sql:
```diff
@@ -62,6 +77,64 @@ create constraint trigger max_users
+    language plpgsql as $$
+    begin
+        if (exists(select 1 from reserved_usernames where name = new.name)) then
+            raise check_violation using message = 'username banned: '||new.name;
```
s/banned/reserved/?

daurnimator reviewed (dismissed):
Looks okay to me other than minor comments left.

RyanSquared commented:
We can drop in a file of reserved usernames and find a way to populate it using the Makefile, right? Should be pretty simple.

KellerFuchs commented:
@RyanSquared Yes, populating a table from a textfile is simple.

RyanSquared committed:
usernames.sql: Copy reserved usernames to table

Fixed by KellerFuchs (the initial commit forgot to add the sql file
to the list in test.sh, resulting in the data not being imported.)

daurnimator commented on usernames.sql:
```diff
@@ -0,0 +1 @@
+copy reserved_usernames (name) from '/tmp/usernames.txt';
```
Why is it in `/tmp`?
Shouldn't it be in e.g. current dir?

RyanSquared replied:
No, this for some reason isn't readable by postgresql. Plus, postgresql requires an absolute path, and copying to /tmp is cleaner.

daurnimator commented on Makefile:
```diff
@@ -22,6 +23,8 @@ install: $(SCHEMA_FILES)
 	$(foreach file,$(SCHEMA_FILES),psql -v ON_ERROR_STOP=1 -h localhost -d userdb -f $(file);)
 
+reserve: usernames.txt usernames.sql
+	psql -v ON_ERROR_STOP=1 -h localhost -d userdb -f usernames.sql
```
Could you add usernames.sql to SCHEMA_FILES so that it's done by `install` like everything else?

KellerFuchs commented:
I merged `master` in to make CI run and solve the merge conflict.
This still needs tests.

RyanSquared commented:
> This still needs tests.

And to be re-PR'd to avoid the disaster of a merge/rebase that happened before

VictorKoenders commented on schema.sql:
```diff
@@ -46,9 +46,25 @@ create table "aux_groups" (
+create function check_hosts_for_hosts() returns trigger
+    language plpgsql as $$
+    begin
+        if ((select count(*) from passwd where passwd.host = new.name) <= new.maxusers) then
```
Should this not be `>= new.maxusers` ?

KellerFuchs replied:
Yeap.
@RyanSquared That PR really needs to include tests.  I will push some for `reserved_usernames` but please do the rest.

KellerFuchs committed:
Document the provenance of the list of reserved names

Also, rename files in preparation for adding other lists.

KellerFuchs committed:
Add names reserved by RFC2142

KellerFuchs committed:
Ignore duplicate or invalid reserved usernames

This avoids having to remove them from imported lists.

KellerFuchs committed:
Makefile: Obey standard Postgre environment variables

KellerFuchs committed:
test.sh: Rely on the Makefile to setup the schema

KellerFuchs commented:
@RyanSquared Not sure what you mean by “re PR'd”

KellerFuchs committed:
Add negative test for `reserved_username`

KellerFuchs reviewed (changes requested):
- [x] I added tests for `reserved_usernames`.
- [ ] Everything else needs tests.
- [ ] `check_hosts_for_hosts` needs fixing; also, it is a silly function name.

KellerFuchs commented:
That's exciting.  :(
Here the check works like a charm.

RyanSquared committed:
schema.sql: Add max_users_on_host constraint

Closes #17

RyanSquared committed:
schema.sql: Disallow colliding user and group names

Closes #28

RyanSquared committed:
schema.sql: Add a `reserved_username` table & matching checks

Closes #27

KellerFuchs commented:
@RyanSquared I did some git surgery on this feature branch, removed commits unintentionally picked from `master` and split your giant initial commit into 3 distinct ones, for the different enforced checks.

KellerFuchs commented:
@RyanSquared Mind checking why the test fails in CI?

RyanSquared commented:
- https://travis-ci.org/hashbang/userdb/builds/382269345#L566 (psql:reserved.sql:15: ERROR:  syntax error at or near "on")
- https://travis-ci.org/hashbang/userdb/builds/382269345#L728 (1. unit_tests.reserved_usernames() --> Successfully inserted invalid user..)

I'll look into the exact details (and how to fix the `on conflict do nothing`) in a bit.

RyanSquared committed:
reserved.sql: Fix `on conflict` not existing in pgsql <9.5

KellerFuchs commented:
@RyanSquared Reminder: you still need to write tests.

RyanSquared committed:
schema.sql: Negate check_hosts_for_hosts check, fix check_max_users check

RyanSquared committed:
tests/06collisions: Add fail-check cases for #17 #28

RyanSquared committed:
tests/06collisions: Add fail-check cases for #27

RyanSquared commented:
@hashbang/administrators ping

benharri commented:
No objections here.

daurnimator reviewed (approved):
I'm happy to merge now; we can always do renames and add more tests later.
This has been blocked for too long.
